components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/nats/StreamingListener.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2020 The Apache Software Foundation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wso2.carbon.inbound.endpoint.protocol.nats;
+
+import io.nats.streaming.StreamingConnection;
+import io.nats.streaming.Subscription;
+import io.nats.streaming.Options;
+import io.nats.streaming.StreamingConnectionFactory;
+import io.nats.streaming.SubscriptionOptions;
+import io.nats.streaming.Message;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.SynapseException;
+import org.wso2.carbon.inbound.endpoint.protocol.nats.management.NatsEndpointManager;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Properties;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Streaming listener class which uses NATS streaming connection to receive messages.
+ */
+public class StreamingListener implements NatsMessageListener {
+
+    private static final Log log = LogFactory.getLog(StreamingListener.class.getName());
+    private String subject;
+    private NatsInjectHandler injectHandler;
+    private Properties natsProperties;
```

Comment:
I see a big similarity between the properties and the behaviour of this class and the CoreListener class. Shall we make NatsMessageListener a super class rather than an interface and move the common behaviours there?

Reply:
The behaviour of StreamingListener and CoreListener is the same but the implementation is different so we cannot do that.